File: Luogu/Tags/莫隊/UVA_12345_Dynamic_len_set_a_L_R.cpp
/*
Mo's algorithm (帶修改莫隊)
除了下標外，其他都和 Luogu P1903 相同
*/
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
const int MX_V = 1e6 + 5;
#define endl '\n'

struct Query {
    int l, r, ts, qid;
    Query(int l, int r, int ts, int qid) : l(l), r(r), ts(ts), qid(qid) {}
};

struct Record {
    int pos, val;
    Record(int pos, int val) : pos(pos), val(val) {}
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    int N, M;
    cin >> N >> M;
    vector<int> A(N);
    for (int i = 0; i < N; i++) cin >> A[i];

    // Mo's algorithm
    LL cur = 0;
    // unordered_map<int, int> cnt;  // TLE
    vector<int> cnt(MX_V);
    auto add = [&](int x) {
        cnt[x]++;
        if (cnt[x] == 1) cur++;
    };
    auto del = [&](int x) {
        cnt[x]--;
        if (cnt[x] == 0) cur--;
    };

    // int BLK_SZ = sqrt(N);
    int BLK_SZ = pow(N, 2.0 / 3.0);
    int qid = 0, ts = 0;  // timestamp
    vector<Query> queries;
    vector<Record> records;
    for (int i = 0; i < M; i++) {
        char op;
        cin >> op;
        if (op == 'Q') {
            int l, r;
            cin >> l >> r;
            r--;
            queries.emplace_back(l, r, ts, qid++);
        } else {
            int pos, val;
            cin >> pos >> val;
            records.emplace_back(pos, val);
            ts++;
        }
    }
    vector<LL> ans(qid, -1);
    sort(queries.begin(), queries.end(), [BLK_SZ](const Query& a, const Query& b) {
        if (a.l / BLK_SZ != b.l / BLK_SZ) return a.l / BLK_SZ < b.l / BLK_SZ;
        if (a.r / BLK_SZ != b.r / BLK_SZ) return a.r / BLK_SZ < b.r / BLK_SZ;
        return a.ts < b.ts;
    });

    int l = 0, r = -1, t = 0;
    for (const auto& q : queries) {
        while (l > q.l) add(A[--l]);
        while (r < q.r) add(A[++r]);
        while (l < q.l) del(A[l++]);
        while (r > q.r) del(A[r--]);
        // 時間戳變大，替換
        while (t < q.ts) {
            auto& [pos, val] = records[t];
            if (l <= pos && pos <= r) {
                del(A[pos]);
                add(val);
            }
            swap(A[pos], val);
            t++;
        }
        // 時間戳變小，還原
        while (t > q.ts) {
            t--;
            auto& [pos, val] = records[t];
            if (l <= pos && pos <= r) {
                del(A[pos]);
                add(val);
            }
            swap(A[pos], val);
        }
        ans[q.qid] = cur;
    }
    for (const auto& x : ans) cout << x << endl;
    return 0;
}

File: Luogu/Tags/莫隊/UVA_12345_Dynamic_len_set_a_L_R.py
import math
from collections import defaultdict

N, M = map(int, input().strip().split())
A = list(map(int, input().strip().split()))

# Mo's algorithm
cnt = defaultdict(int)
cur = 0
def add(x: int) -> None:
    global cur
    cnt[x] += 1
    if cnt[x] == 1:
        cur += 1

def delete(x: int) -> None:
    global cur
    cnt[x] -= 1
    if cnt[x] == 0:
        cur -= 1

BLK_SZ = math.pow(N, 2.0 / 3.0)

qid = ts = 0
queries = []  # (l, r, ts, qid)
records = []  # (pos, val)
for _ in range(M):
    op, *args = input().split()
    if op == 'Q':
        l, r = map(int, args)
        r -= 1  # [l, r]
        queries.append([l, r, ts, qid])
        qid += 1
    else:
        pos, val = map(int, args)
        records.append([pos, val])
        ts += 1

ans = [-1] * len(queries)
queries.sort(key=lambda x: (x[0] // BLK_SZ, x[1] // BLK_SZ, x[2]))
l, r, t = 0, -1, 0
for ql, qr, ts, qid in queries:
    while l > ql:
        l -= 1
        add(A[l])
    while r < qr:
        r += 1
        add(A[r])
    while l < ql:
        delete(A[l])
        l += 1
    while r > qr:
        delete(A[r])
        r -= 1
    while t < ts:
        pos, val = records[t]
        if l <= pos <= r:
            delete(A[pos])
            add(val)
        A[pos], records[t][1] = records[t][1], A[pos]
        t += 1
    while t > ts:
        t -= 1
        pos, val = records[t]
        if l <= pos <= r:
            delete(A[pos])
            add(val)
        A[pos], records[t][1] = records[t][1], A[pos]
    ans[qid] = cur

print(*ans, sep='\n')